### PR TITLE
File Upload Over HTTP

### LINKED ISSUE #56: Contributing File Upload Support
I'm looking to contribute support for IoT Hub file uploads as per https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-file-upload. The purpose of this issue is to signal that it's being worked on and to get the conversation going regarding the feature.

Is there anything I should take into account when implementing the feature?

### PR DESCRIPTION
**Pull request by alexg-axis**
Implement File Upload over HTTP according to https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/iot-hub/iot-hub-devguide-file-upload.md#device-initialize-a-file-upload.

Device:
* Implements a new HTTP transport with support for the file-related APIs
  * Initialize a file upload
  * Upload file using Azure storage APIs
  * Notify IoT Hub of a completed file upload

Service:
* Implements proper support for watching for file notifications 

Closes #56 

To test:

Set up an Azure IoTHub and an Azure Storage Account. Configure File Upload for the IoTHub to use the Storage Account.

```sh
go build -o iot-device cmd/iothub-device/main.go
go build -o iot-service cmd/iothub-service/main.go
```

Then, in one terminal:
```sh
export IOTHUB_SERVICE_CONNECTION_STRING="xxx"
./iot-service watch-file-notifications
```

In another terminal, upload a file:
```sh
export IOTHUB_DEVICE_CONNECTION_STRING="xxx"
./iot-device -transport http file-upload ./path/to/file.txt
```

After a few seconds the file notification will be printed.

```json
{
	"deviceId": "my-device",
	"blobUri": "https://my-blob.blob.core.windows.net/iothub/my-device/file.txt",
	"blobName": "file.txt",
	"lastUpdatedTime": "2021-11-15T12:50:31Z",
	"blobSizeInBytes": 6,
	"enqueuedTimeUtc": "2021-11-15T12:50:32.3018639Z"
}
```

### CODE REVIEW

**Comment by alexg-axis** [comment]
Resources:

* https://github.com/MicrosoftDocs/azure-docs/blob/master/articles/iot-hub/iot-hub-devguide-file-upload.md#device-initialize-a-file-upload
* https://docs.microsoft.com/en-us/rest/api/storageservices/create-service-sas
* https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-node-node-file-upload

File: iotdevice/transport/http/http.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+func (tr *Transport) GetBlobSharedAccessSignature(ctx context.Context, blobName string) (string, string, error) {
+	payload := CreateFileUploadRequest{
+		BlobName: blobName,
+	}
+	body, err := json.Marshal(&payload)
+	if err != nil {
+		return "", "", err
+	}
+
+	req, err := http.NewRequest(http.MethodPost, fmt.Sprintf("https://localhost:443/devices/%s/files", tr.creds.GetDeviceID()), bytes.NewReader(body))
```
Comment (amenzhinsky) [inline iotdevice/transport/http/http.go:98]: What's supposed to be running on localhost?

File: iotdevice/transport/http/http.go [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+	req, err := http.NewRequest(http.MethodPut, sasURI, file)
+	if err != nil {
+		return err
+	}
+
+	_, err = tr.client.Do(req)
+	if err != nil {
+		return err
+	}
+
```
Comment (amenzhinsky) [inline iotdevice/transport/http/http.go:128]: Need to handle response somehow and close res.Body

**Comment by amenzhinsky** [comment]
Thanks for the PR, in general everything looks good, just a few thins to polish

File: iotdevice/transport/http/http.go (hunk 1 above)
Comment (alexg-axis) [reply iotdevice/transport/http/http.go:98]: I'm running it towards a local mocked version of Azure's IoTHub, it's still work in progress, hence the hard-coded URL.

**Comment by amenzhinsky** [comment]
Thanks for the contribution

**Pull request merged**